DSC

diegonieto opened the pull request.

diegonieto committed:
dsc: add dscsigner

diegonieto committed:
dsc: add dscverifier element

diegonieto committed:
dsc: add dsc tests

ylatuya commented:
- The signer works with video segments instead of video frames; the duration of the segments must be configurable.
- The signature spans over the hash of all video data packets and the hash of the previous segment.  It should also include the content UUID, the hash method, and start and end flags are included in the signature, so that this data cannot be modified. (See IBC2025PaperAuthentication_HHI.pdf and probably the spec to understand how it's done)
- The Signature meta must include all the necessary information to generate the DSCI SEI message (hash method, public key, etc...)
- The verifier should not need the public key information; the public key is part of the DSCI SEI message, and it should be available in the meta.

diegonieto committed:
dsc: move to video folder

diegonieto committed:
dsc: sign and verify raw frames within a gop instead of sign frame by frame.

Use GOP-1 signing approach
Hash both previous digest and current one

diegonieto committed:
dsc: hash only the selected NALUs instead of the whole bitstream

diegonieto committed:
dsc: use metadata cert uri instead of verifier one

Compose the public key uri based on key-storate-path in verifier and
cert_uri metadata

diegonieto committed:
dsc: format fix

diegonieto committed:
dsc: signaturemeta format fix

diegonieto committed:
dsc: nal_parser format fix

diegonieto committed:
dsc: add module descriptions

diegonieto committed:
dsc: update signature meta test

diegonieto committed:
dsc: update nal parser test

diegonieto committed:
dsc: fix dsc_substream format

diegonieto committed:
dsc: update dsc test

diegonieto committed:
dsc: update README.md

diegonieto committed:
dsc: update cargo

diegonieto committed:
dsc: update plugin description

diegonieto committed:
dsc: update commit README.md

diegonieto committed:
dsc: add dscinserter skeleton

diegonieto committed:
dsc: wip dscinserter

diegonieto committed:
dscverifier: make dscverifier compatible with VVC/VTM one

diegonieto committed:
dscverifier: polish implementation

diegonieto committed:
dscsigner: make dscsigner compatible with VVC/VTM one

ylatuya commented on video/dsc/src/verifier/imp.rs:
```diff
@@ -0,0 +1,609 @@
+        let selection_meta = buffer.meta::<gst_video::video_meta::VideoDSCSelectionMeta>();
+
+        if let Some(init_meta) = initialization_meta {
+            let dsc_init = init_meta.dsc_initialization();
```
This function is too big. Create a function to handle each of the metas.

diegonieto replied:
refactored :+1:

ylatuya commented on video/dsc/src/verifier/imp.rs:
```diff
@@ -0,0 +1,609 @@
+            let cert_uri = gop_state.current_cert_uri.clone();
+            
+            let pkey = if let Some(uri) = cert_uri {
+                match self.load_public_key_from_cert_uri(&uri, &mut gop_state) {
```
Is the public key likely to change? It's maybe worth caching it.

ylatuya replied:
I see that load_public_key_from_cert_uri already caches keys

diegonieto replied:
yes: `gop_state.public_key_cache.insert(cert_uri.to_string(), pkey.clone());`

ylatuya commented on video/dsc/src/verifier/imp.rs:
```diff
@@ -0,0 +1,609 @@
+                return Err(gst::FlowError::Error);
+            };
+
+            if let Some(ref mut dsc_manager) = gop_state.dsc_manager {
```
Factor out the code into functions for these parts too, it's otherwise very hard to follow the code

diegonieto replied:
updated :+1:

ylatuya commented on video/dsc/src/verifier/imp.rs:
```diff
@@ -0,0 +1,609 @@
+                            }
+                        };
+
+                        gst::debug!(*CAT, imp = self, "VERIFIER: Data packet content (first 32 bytes): {:02x?}", &data_packet[..std::cmp::min(32, data_packet.len())]);
```
These logs probably be verbose or trace.

diegonieto replied:
moving to "log" instead

ylatuya commented on video/dsc/src/verifier/imp.rs:
```diff
@@ -0,0 +1,609 @@
+                            }
+                        };
+
+                        gst::debug!(*CAT, imp = self, "VERIFIER: Data packet content (first 32 bytes): {:02x?}", &data_packet[..std::cmp::min(32, data_packet.len())]);
```
Why the VERIFIER:  prefix?  This will all be log under the defined category for the element

diegonieto replied:
I used capital prefixed to debug easier, but now it needs to be removed, yes :+1:

ylatuya commented on video/dsc/src/verifier/imp.rs:
```diff
@@ -0,0 +1,609 @@
+                        return Err(gst::FlowError::Error);
+                    }
+
+                    gst::info!(*CAT, imp = self, "VERIFIER_TRACE: Added NAL to substream {}, size: {}, first 32: {:02x?}, last 32: {:02x?}",
```
These traces needs to be cleaned

diegonieto replied:
done

ylatuya commented on video/dsc/src/verifier/imp.rs:
```diff
@@ -0,0 +1,609 @@
+                            Ok(true) => {
+                                gst::info!(*CAT, imp = self, "✅ Substream signature verified successfully");
+                            },
+                            Ok(false) => {
```
Verification errors shouldn't finish processing of the stream or it should at least be configurable through a property, so that users can define if a verification error causes an error or not.
In some scenarios, the application will decide what to do during verification. This can be done with a signal.

diegonieto replied:
Updated. I've added "fail-on-verification-error" property to handle that. Now verification outcome is always posted in the bus, so applications can be notified as well.

ylatuya commented:
Commits should also be squashed to upstream it.

ylatuya commented on video/dsc/Cargo.toml:
```diff
@@ -0,0 +1,43 @@
+gst-video = { workspace = true, features = ["v1_30"] }
+human_bytes = { version = "0.4", default-features = false }
+atomic_refcell = "0.1"
+gst-plugin-originalbuffer = { git = "https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs", tag = "0.13.0" }
```
Is this dependency needed?

diegonieto replied:
It is not, removed. Thanks!

ylatuya commented on video/dsc/README.md:
```diff
@@ -0,0 +1,19 @@
+# GStreamer DSC Plugin
+
+A GStreamer plugin for Digital Signed Content (DSC) that provides cryptographic signing and verification for generic data based on OpenSSL signing and verification.
```
- This should have a reference to the spec
- Is `generic data` appropriate here? Only certain video codecs are supported, which should be listed here as well.
- OpenSSL is a library, not the cryptographic method used for signing

diegonieto replied:
updated :+1:

ylatuya commented on video/dsc/README.md:
```diff
@@ -0,0 +1,19 @@
+
+A GStreamer plugin for Digital Signed Content (DSC) that provides cryptographic signing and verification for generic data based on OpenSSL signing and verification.
+
+## Features
```
It's plugins, not features

diegonieto replied:
updated :+1:

ylatuya commented on video/dsc/src/nal_parser.rs:
```diff
@@ -0,0 +1,243 @@
+
+static CAT: LazyLock<gst::DebugCategory> = LazyLock::new(|| {
+    gst::DebugCategory::new(
+        "nal-parser",
```
I would use `dsc-nal-parser` so that's easier to filter logs with the `dsc*` prefix

diegonieto replied:
updated :+1:

ylatuya commented on video/dsc/src/nal_parser.rs:
```diff
@@ -0,0 +1,243 @@
+            start = end;
+        }
+
+        gst::info!(*CAT, "NAL_TRACE: Total signable data: {} NAL units extracted", nal_units.len());
```
Cleanup `NAL_TRACE` from debug logs

diegonieto committed:
dscverifier: clean traces

diegonieto committed:
dsc: remove dependency not needed

diegonieto committed:
dsc: update dsc nal parser debug name

diegonieto committed:
dsc: dsc nal parser. Update traces

diegonieto committed:
dscverifier: refactor transform_ip

diegonieto committed:
dscverifier: refactor verification process

diegonieto committed:
dscverifier: post a message with verification result

Avoid to fail on verification by default. Use the new
"fail-on-verification-error" property for it.

diegonieto committed:
dsc: update README.md

diegonieto committed:
dsc: update tests

diegonieto committed:
dsc: update dsc tests
